accepted/2020/workloads/workload-manifest.md

```diff
@@ -0,0 +1,430 @@
+# .NET SDK Workload Manifests
+
+# Overview
```

Comment:
This is a little off-topic, and I don't have a better place for these feedback, so choosing here.

We've been having some conversations about adopting how to declare workloads that are not explicitly tied to a TFM. ASP.NET Core and Windows Forms are good examples. We've been discussing three models to declare a dependence on such a workload in a project file (using ASP.NET Core as the example):

* `UserAspnetCore`
* `<Workloads>AspnetCore</Workloads>`
* `Sdk="Microsoft.NET.Sdk.Web"`

It is worth clarifying if these are all equivalent from the perspective of tying a project file to a workload, or if there are semantic or technical differences.

If anyone has a preference for the scheme they would like to see used, independent from my question, please feel free to weigh in.

Reply:
Since you asked for people to weigh in, here are a bunch of my thoughts:

As a user, I would strongly prefer that Sdk="Microsoft.NET.Sdk.Web" continue to work in .NET 5, so as to have one less thing to need to change. It obviously needs to be able to work for projects targeting 3.x, since if installing .NET 5 SDK broke building 3.x apps unless you use global.json, people would be upset.

Furthermore, I'm not certain that we want every SDK the users might need to specify to become a unique workload. For example, consider `Microsoft.NET.Sdk.Worker`. It sets some capabilities, some properties, and enables copying `*.config` and `*.json` to the output (since M.E.Hosting applications are likely to want to use appsetting.json). Should that become a separate workload? Realistically it could and probably should just be bundled with whatever lets you build simple console apps, which would surely be included in the default .NET 5 SDK install.

Now perhaps that particular SDK is killed off in favor of incorporating that logic into the core SDK, activated by a property or something. But similar things could obviously exist. Not to mention that third party frameworks will still exist that will have their own `Sdk=` needed, unless there is a plan to support for third parties creating their own workloads. (Well, presumably they could, but without a pre-installed manifest, they would not get the desired experience.)

Per the resolver spec, the missing workloads are already discovered by a new SDK Resolver when an SDK being imported is found to not be present. I'd expect that would work just as well being triggered by the project file itself. 

Granted, that this does complicate resolving. Knowing which SDK pack is needed would not always uniquely resolve to a single workload. How `dotnet bundle restore` would pick between them is unclear. One advantage of `<Workloads>AspnetCore</Workloads>` is the ability to explicitly specify the workload by name, which could be used to override the the default workload selection logic if you really wanted something different than it decided. Presumably, something like "dotnet bundle restore --build-only" would be smart enough to see the specified pack is a "dev" one, and look though its extends, find only the build workloads, and install just the packs those need.

Reply:
Great feedback. Thanks for taking the time.

First, we don't plan to break the existing syntax. That will continue to work. This conversation is a question of what the new scheme should be.

It's subtle but the topic at hand doesn't really relate to workloads exactly. For example, we already have and plan to continue to use `UseWinforms` and `UseWPF` properties. We're still planning to include these two components together in the Windows desktop workload, which will be declared via the `net5.0-windows` TFM.

It's more complicated with ASP.NET Core since it doesn't have its own TFM. ASP.NET Core will be available within `net5.0`, however, we don't want it to be required to run any .NET 5.0 app. As a result, we need a specific opt-in gesture. In short, it either can work the new way (with the `UseFoo` style syntax with the benefit that ASP.NET and WPF, for example, will have a symmetrical experience, or ASP.NET can be different and work the old way.

We talked to the ASP.NET team. They didn't like either option, for opposite reasons. I suspect they will end up going with the `UseFoo` style syntax so that the platform looks like same for all app types.

Make sense?

Reply:
Admittedly your first post in this comment series did not distinguish that WPF would be part of a windows workload, and not a separate one. I’ve not actually used the WindowsDesktop sdk, so had not noticed that it used only properties instead of SDKs to distinguish Winforms and WPF. (Admittedly, since using both in the same project is possible, properties are more ergonomic than using sdks, since sdks would require adding a stand alone sdk element to the file to specify the second one if both were used). 

My main objection to `UseFoo` for scenarios like ASP.NET Core, is that SDKs are already a user exposed concept, so adding a magic property to enable one merely for consistency, just seems to be adding more user exposed concepts. After all, third party SDKs will realistically not be able to take advantage of magic properties to load themselves, but could use Sdk= to do so. (Yeah, technically, there are ways a third party sdk could use a magic property to do this, but only if already installed via an installer/package-manager. It could not benefit from automatic discovery that the sdk is needed but not installed. And third party SDKs hooking themselves into every single build on the system is probably not something you would want to encourage, given how easily doing so could cause weird breakage.)